docs/topics/media-pipeline.rst

```diff
@@ -392,6 +392,30 @@ See here the methods that you can override in your custom Files Pipeline:
 
 .. class:: FilesPipeline
 
+   .. method:: file_path(request, response, info)
+
+      This method is called once per downloaded item. It returns the
+      download path of the file originating from the specified
+      :class:`response <scrapy.http.Response>`.
+
+      In addition to ``response``, this method receives the original
+      :class:`request <scrapy.Request>` and
+      :class:`info <scrapy.pipelines.media.MediaPipeline.SpiderInfo>`.
+
+      You can override this method to customize the download path of each file.
+
+      For example, to download all files into the ``files`` folder with their
+      original filenames::
```

Comment:
I think we should be a bit more clear on that it is just an example, which works for a certain kind of URLs - `http://example.com/foo.png` would be stored properly, but e.g. all files like `http://example.com/file?name=foo.png` would be stored as "file", overwriting each other. Possible solutions:

* say that it is an example, which works on websites where file URLs follow certain rules
* make it more robust (but less readable) by adding full URL hash to the path, in addition to guessed file name.

A full implementation should probably also check Content-Disposition header, to get a file name.

Reply:
I’ve gone with option 1, and also tried to make it obvious that the example implementation does not take subdirectories into account either.